Update doc-Refitting NumPyro models with ArviZ (and xarray) #1801

Divyateja04 opened the pull request:
## Description
<!--
Thank you so much for your PR! To help us review your contribution, please consider the following points:

- The PR title should summarize the changes, for example "Add new group argument for the pair plot".
  Avoid non-descriptive titles such as "Addresses issue #348".

- The description should provide at least 1-2 sentences describing the pull request in detail and
  link to any relevant issues.

- Please prefix the title of incomplete contributions with [WIP] (to indicate a work in
  progress). WIPs may be useful to (1) indicate you are working on something to avoid
  duplicated work, (2) request broad review of functionality or API, or (3) seek collaborators. -->

## Checklist
<!-- Feel free to remove check-list items aren't relevant to your change -->

- [x] Follows [official](https://github.com/arviz-devs/arviz/blob/main/CONTRIBUTING.md#pull-request-checklist) PR format
- [ ] Includes a sample plot to visually illustrate the changes (only for plot-related functions)
- [x] New features are properly documented (with an example if appropriate)?
- [ ] Includes new or updated tests to cover the new feature
- [x] Code style correct (follows pylint and black guidelines)
- [ ] Changes are listed in [changelog](https://github.com/arviz-devs/arviz/blob/main/CHANGELOG.md#v0xx-unreleased)

<!--
Also, please consider reading the contributing guidelines and code of conduct carefully before submitting the PR. They are available at
- https://github.com/arviz-devs/arviz/blob/main/CONTRIBUTING.md
- https://github.com/arviz-devs/arviz/blob/main/CODE_OF_CONDUCT.md

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in. Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->
Changes Done:
Replaced `az.reloo` with {func}`~arviz.reloo`
Edited some grammatical mistakes
Replaced PyStanSamplingWrapper terms with NumPyroSamplingWrapper
Replaced "Stan Code" with "NumPyro Code"
Replaced pystan_wrapper to numpyro_wrapper

roshnaeem commented:
Thank you so much for the PR. Can you please make the following changes?
1. Stan is a probabilistic programming language for statistical inference, so you don't need to replace `Stan` with `NumPyro`. 
2. Here, we are talking about `PyStanSamplingWrapper`, you might have checked by going to the [link](https://arviz-devs.github.io/arviz/api/generated/arviz.PyStanSamplingWrapper.html). In all the places where you have changed the `pystan_wrapper` to `numpyro_wrapper`, there should be `pystan_wrapper`. `numpyro_wrapper` link does not work, you can check by going to the link.

roshnaeem commented:
@Divyateja04 , you can check the preview of the changes you have made by following the given steps:
1. Go to the checks of your PR. The last one is `docs/readthedocs.org:arviz`.
2. IF this check has completed running and is successfully build(green tick), click on the `Details` button in front of it. 
3. It will open the docs, go to the webpage you made changes, and check the changes you have made to that file. 

This way you will be able to check the changes and test the links that you have changed.

Divyateja04 commented:
Yeah I replaced Numpyro with Stan where required and changed the wrappers too. In the PyStan issue code, there's a class `PyStan2SamplingWrapper` but I couldn't find this in code and only found `PyStanSamplingWrapper` so wanted to know if that 2 should be there or not

Divyateja04 commented:
And if you don't mind me asking why are the first two tests failing?

Divyateja04 commented:
And also are there any other things I need to change and once again sorry for the trouble!

Divyateja04 commented:
Whenever you are free, can you check and let me know?

roshnaeem commented:
yes, there is [arviz.PyStan2SamplingWrapper](https://arviz-devs.github.io/arviz/api/generated/arviz.PyStan2SamplingWrapper.html). You don't need to make any changes in the code. Keep the code as it is. 
Yeah, you are right, the changes should be merged, we are working on it. Also, let me check the test failing issues.

roshnaeem commented on doc/source/user_guide/numpyro_refitting_xr_lik.ipynb:
```diff
@@ -4537,9 +4538,13 @@
    "metadata": {},
    "source": [
-    "We will create a subclass of `az.SamplingWrapper`. Therefore, instead of having to implement all functions required by `az.reloo` we only have to implement `sel_observations` (we are cloning `sample` and `get_inference_data` from the `PyStanSamplingWrapper` in order to use `apply_ufunc` instead of assuming the log likelihood is calculated within Stan). \n",
+    "We will create a subclass of {class}`~arviz.SamplingWrapper`. Therefore, instead of having to implement all functions required by {func}`~arviz.reloo` we only have to implement {func}`~arviz.SamplingWrapper.sel_observations` (we are cloning {func}`~arviz.PyStanSamplingWrapper.sample` and {func}`~arviz.PyStanSamplingWrapper.get_inference_data` from the {class}`~arviz.PyStan2SamplingWrapper` i ...
```
"instead of assuming the log likelihood is calculated within Stan." Here, we are talking Stan. Can you please change it back to Stan.

Divyateja04 replied:
Done

roshnaeem commented on doc/source/user_guide/numpyro_refitting_xr_lik.ipynb:
```diff
@@ -2090,11 +2091,11 @@
    "metadata": {},
    "source": [
-    "We are now missing the `log_likelihood` group because we have not used the `log_likelihood` argument in `idata_kwargs`. We are doing this to ease the job of the sampling wrapper. Instead of going out of our way to get Stan to calculate the pointwise log likelihood values for each refit and for the excluded observation at every refit, we will compromise and manually write a function to calcul ...
+    "We are now missing the `log_likelihood` group because we have not used the `log_likelihood` argument in `idata_kwargs`. We are doing this to ease the job of the sampling wrapper. Instead of going out of our way to get NumPyro to calculate the pointwise log likelihood values for each refit and for the excluded observation at every refit, we will compromise and manually write a function to cal ...
```
Here it will be "Instead of going out of our way to get Stan to calculate the pointwise log likelihood values for each refit ". The Stan will remain as it is.

Divyateja04 replied:
Done

roshnaeem commented on doc/source/user_guide/numpyro_refitting_xr_lik.ipynb:
```diff
@@ -2090,11 +2091,11 @@
-    "We are now missing the `log_likelihood` group because we have not used the `log_likelihood` argument in `idata_kwargs`. We are doing this to ease the job of the sampling wrapper. Instead of going out of our way to get Stan to calculate the pointwise log likelihood values for each refit and for the excluded observation at every refit, we will compromise and manually write a function to calcul ...
+    "We are now missing the `log_likelihood` group because we have not used the `log_likelihood` argument in `idata_kwargs`. We are doing this to ease the job of the sampling wrapper. Instead of going out of our way to get NumPyro to calculate the pointwise log likelihood values for each refit and for the excluded observation at every refit, we will compromise and manually write a function to cal ...
     "\n",
-    "Even though it is not ideal to lose part of the straight out of the box capabilities of PyStan-ArviZ integration, this should generally not be a problem. We are basically moving the pointwise log likelihood calculation from the Stan code to the Python code, in both cases we need to manually write the function to calculate the pointwise log likelihood.\n",
```
The same goes for `PyStan-ArviZ integration`.

Divyateja04 replied:
Done but actually I never changed any code but rather the docs. I mean onlly the text content was changed, so all the test should still run right?(just asking)

roshnaeem replied:
Yeah, there is some other issue with tests, nothing is wrong with your PR. 
Thank you so much for your PR. I will merge it as soon as the other gets merged. Feel free to explore other issues as well.

Divyateja04 replied:
So um, I didn't do anything wrong with this PR right and I'm good to go?(I'm mean was everything I did correct?)

roshnaeem replied:
Yes. Thank you :))

Divyateja04 commented:
Hey, by the way since this hasn't been merged yet, can you merge it in october so that it counts for haktoberfest
Also are there any other docs that need to be updated that I can work on?

roshnaeem commented:
@Divyateja04 yeah, we are working on more issues of documentation for beginners. We'll create plenty of issues in the coming weeks. Please keep checking. 
Yeah, I can merge it in October. Otherwise you can always work on the new issues in October.

Divyateja04 commented:
yeah i'm planning to work on issues too..like I have some knowledge about python but not that much in depth so asked what else I can work on

roshnaeem commented:
@Divyateja04, yes you can work on documentation and we are here to help.

roshnaeem commented:
Hello @Divyateja04, can you do me a favor? The related PR  https://github.com/arviz-devs/arviz/pull/1798 has been merged, there were few more changes I did it in, so can you please check again and do one more commit to keep your PR the same as #1798 is. 

Alos, i have opened new issues for hacktoberfest, you can start working on them.

Divyateja04 commented:
Yeah sure, lemme check and get back to you

Divyateja04 commented:
I updated it

roshnaeem commented on doc/source/user_guide/numpyro_refitting_xr_lik.ipynb:
```diff
@@ -4632,15 +4637,15 @@
     "We use `idata_orig` as a starting point, and mostly as a source of observed and constant data which is then subsetted in `sel_observations`.\n",
     "\n",
-    "Finally, `sample_kwargs` and `idata_kwargs` are used to make sure all refits and corresponding InferenceData are generated with the same properties."
+    "Finally, `sample_kwargs` and `idata_kwargs` are used to make sure all refits and corresponding to `InferenceData` are generated with the same properties."
```
corresponding `InferenceData`

roshnaeem commented on doc/source/user_guide/numpyro_refitting_xr_lik.ipynb:
```diff
@@ -13,7 +13,7 @@
    "metadata": {},
    "source": [
-    "Below there is one example of `SamplingWrapper` usage for NumPyro."
+    "Below there is an example of `SamplingWrapper` usage for NumPyro."
```
Can you please add the link to NumPyro docs? 
"[NumPyro] (https://pyro.ai/numpyro/)"

roshnaeem commented on doc/source/user_guide/numpyro_refitting_xr_lik.ipynb:
```diff
@@ -141,7 +141,8 @@
    "source": [
-    "We have defined a dictionary `sample_kwargs` that will be passed to the `SamplingWrapper` in order to make sure that all refits use the same sampler parameters. We follow the same pattern with `az.from_numpyro`."
+    "We have defined a dictionary `sample_kwargs` that will be passed to the `SamplingWrapper` in order to make sure that all\n",
+    "refits use the same sampler parameters. We follow the same pattern with {func}`~az.from_numpyro`."
```
Please change {func}`~az.from_numpyro` to {func}`az.from_numpyro <arviz.from_numpyro>`

OriolAbril replied:
using `` {func}`~arviz.from_numpyro` `` or `` {func}`arviz.from_numpyro` `` is perfectly good enough, no need to use custom text if it's only for the `az.` part. It is important to change what is used now though because sphinx doesn't use alias when  importing, so it won't generate a clickable link if using `az.from_numpyro` as reference.

Divyateja04 replied:
So um should I change anything or is it good to go?

OriolAbril replied:
it is already good to go, just wanted to comment to ease future work

Divyateja04 replied:
oh okay sure, can you tell me when it'll be merged though

Divyateja04 commented:
yeah really sorry, I actually thought the only change was in the first part....really sorry...updating it

roshnaeem commented on doc/source/user_guide/numpyro_refitting_xr_lik.ipynb:
```diff
@@ -13,7 +13,7 @@
    "metadata": {},
    "source": [
-    "Below there is one example of `SamplingWrapper` usage for NumPyro."
+    "Below there is an example of `SamplingWrapper` usage for [NumPyro] (https://pyro.ai/numpyro/)."
```
remove the space between [](). I added it because the markdown was rendering it to link.

roshnaeem commented on doc/source/user_guide/numpyro_refitting_xr_lik.ipynb:
```diff
@@ -141,7 +141,8 @@
    "source": [
-    "We have defined a dictionary `sample_kwargs` that will be passed to the `SamplingWrapper` in order to make sure that all refits use the same sampler parameters. We follow the same pattern with `az.from_numpyro`."
+    "We have defined a dictionary `sample_kwargs` that will be passed to the `SamplingWrapper` in order to make sure that all\n",
+    "refits use the same sampler parameters. We follow the same pattern with {func}`~az.from_numpyro <arviz.from_numpyro>`."
```
Remove the `~` from {func}`~az.from_numpyro <arviz.from_numpyro>`

Divyateja04 replied:
that space is still there?

roshnaeem commented on doc/source/user_guide/numpyro_refitting_xr_lik.ipynb:
```diff
@@ -4537,9 +4538,13 @@
    "metadata": {},
    "source": [
-    "We will create a subclass of `az.SamplingWrapper`. Therefore, instead of having to implement all functions required by `az.reloo` we only have to implement `sel_observations` (we are cloning `sample` and `get_inference_data` from the `PyStanSamplingWrapper` in order to use `apply_ufunc` instead of assuming the log likelihood is calculated within Stan). \n",
+    "We will create a subclass of {class}`~arviz.SamplingWrapper`. Therefore, instead of having to implement all functions required by {func}`~arviz.reloo` we only have to implement {func}`~arviz.SamplingWrapper.sel_observations` (we are cloning {func}`~arviz.PyStanSamplingWrapper.sample` and {func}`~arviz.SamplingWrapper.get_inference_data` from the {class}`~arviz.SamplingWrapper` in order to us ...
```
{func}`~arviz.PyStanSamplingWrapper.sample` to {func}`~arviz.SamplingWrapper.sample`

Divyateja04 replied:
Changed them

Divyateja04 commented:
Hey, can you merge this so that I can work on the other issue, otherwise I get really weird PR issues when trying to PR when one active PR is already there

roshnaeem commented:
It'll be merged soon. Make a new git branch for other changes.

Divyateja04 commented:
sure, just wanted to check with you though

roshnaeem commented:
You did not need to push your changes in this branch. It was okay, you needed to make a new branch and do all the steps in the new branch.

Divyateja04 commented:
Yeah I made a new branch and made the changes for the other PR, can you check?

Divyateja04 commented:
Just a small query though, should I add changes made for this issue too in that PR?

roshnaeem commented:
It'll only include plot_posterior. and this will only include `numpyro_refitting_xr_lik.ipynb`. every branch should have separate changes so the things don't get mixed up.  The other PR is right.

Divyateja04 commented:
Okay so you want me to create a new branch, make the changes and send it again?

roshnaeem commented:
yes, you can do it.

Divyateja04 commented:
I'll just name it numpyro-xarray
